// vso-task-lib/vsotask.js
/// <reference path="../definitions/node.d.ts" />
/// <reference path="../definitions/Q.d.ts" />
/// <reference path="../definitions/shelljs.d.ts" />
/// <reference path="../definitions/minimatch.d.ts" />
/// <reference path="../definitions/glob.d.ts" />
"use strict";
var shell = require('shelljs');
var fs = require('fs');
var path = require('path');
var os = require('os');
var minimatch = require('minimatch');
var globm = require('glob');
var tcm = require('./taskcommand');
var trm = require('./toolrunner');
(function (TaskResult) {
    TaskResult[TaskResult["Succeeded"] = 0] = "Succeeded";
    TaskResult[TaskResult["Failed"] = 1] = "Failed";
})(exports.TaskResult || (exports.TaskResult = {}));
var TaskResult = exports.TaskResult;
//-----------------------------------------------------
// String convenience
//-----------------------------------------------------
function startsWith(str, start) {
    return str.slice(0, start.length) == start;
}
function endsWith(str, end) {
    return str.slice(-str.length) == end;
}
//-----------------------------------------------------
// General Helpers
//-----------------------------------------------------
exports._outStream = process.stdout;
exports._errStream = process.stderr;
function _writeError(str) {
    exports._errStream.write(str + os.EOL);
}
exports._writeError = _writeError;
function _writeLine(str) {
    exports._outStream.write(str + os.EOL);
}
exports._writeLine = _writeLine;
function setStdStream(stdStream) {
    exports._outStream = stdStream;
}
exports.setStdStream = setStdStream;
function setErrStream(errStream) {
    exports._errStream = errStream;
}
exports.setErrStream = setErrStream;
//-----------------------------------------------------
// Results and Exiting
//-----------------------------------------------------
function setResult(result, message) {
    debug('task result: ' + TaskResult[result]);
    command('task.complete', { 'result': TaskResult[result] }, message);
    if (result == TaskResult.Failed) {
        _writeError(message);
    }
    if (result == TaskResult.Failed && !process.env['TASKLIB_INPROC_UNITS']) {
        process.exit(0);
    }
}
exports.setResult = setResult;
function handlerError(errMsg, continueOnError) {
    if (continueOnError) {
        error(errMsg);
    }
    else {
        setResult(TaskResult.Failed, errMsg);
    }
}
exports.handlerError = handlerError;
//
// Catching all exceptions
//
process.on('uncaughtException', function (err) {
    setResult(TaskResult.Failed, 'Unhandled:' + err.message);
});
function exitOnCodeIf(code, condition) {
    if (condition) {
        setResult(TaskResult.Failed, 'failure return code: ' + code);
    }
}
exports.exitOnCodeIf = exitOnCodeIf;
//
// back compat: should use setResult
//
function exit(code) {
    setResult(code, 'return code: ' + code);
}
exports.exit = exit;
//-----------------------------------------------------
// Loc Helpers
//-----------------------------------------------------
var locStrings = {};
var resourceFile;
function setResourcePath(path) {
    checkPath(path, 'resource file path');
    resourceFile = path;
    debug('set resource file to: ' + resourceFile);
    var resourceJson;
    try {
        resourceJson = require(resourceFile);
    }
    catch (err) {
        setResult(TaskResult.Failed, 'Invalid json in resource file: ' + path); // exit
    }
    if (resourceJson && resourceJson.hasOwnProperty('messages')) {
        debug('cache loc strings');
        for (var key in resourceJson.messages) {
            locStrings[key] = resourceJson.messages[key];
        }
    }
    else {
        warning('there is no messages section in resource file: ' + resourceFile);
    }
}
exports.setResourcePath = setResourcePath;
function loc(key, defaultStr) {
    if (!resourceFile) {
        warning('resource file haven\'t been set, retrun default loc string.');
        return defaultStr;
    }
    if (locStrings.hasOwnProperty(key)) {
        return locStrings[key];
    }
    else {
        return defaultStr;
    }
}
exports.loc = loc;
//-----------------------------------------------------
// Input Helpers
//-----------------------------------------------------
function getVariable(name) {
    var varval = process.env[name.replace(/\./g, '_').toUpperCase()];
    debug(name + '=' + varval);
    return varval;
}
exports.getVariable = getVariable;
function setVariable(name, val) {
    if (!name) {
        _writeError('name required: ' + name);
        exit(1);
    }
    var varValue = val || '';
    process.env[name.replace(/\./g, '_').toUpperCase()] = varValue;
    debug('set ' + name + '=' + varValue);
    command('task.setvariable', { 'variable': name || '' }, varValue);
}
exports.setVariable = setVariable;
function getInput(name, required) {
    var inval = process.env['INPUT_' + name.replace(' ', '_').toUpperCase()];
    if (inval) {
        inval = inval.trim();
    }
    if (required && !inval) {
        setResult(TaskResult.Failed, 'Input required: ' + name);
    }
    debug(name + '=' + inval);
    return inval;
}
exports.getInput = getInput;
function getBoolInput(name, required) {
    return getInput(name, required) == "true";
}
exports.getBoolInput = getBoolInput;
function setEnvVar(name, val) {
    if (val) {
        process.env[name] = val;
    }
}
exports.setEnvVar = setEnvVar;
//
// Split - do not use for splitting args!  Instead use arg() - it will split and handle
//         this is for splitting a simple list of items like targets
//
function getDelimitedInput(name, delim, required) {
    var inval = getInput(name, required);
    if (!inval) {
        return [];
    }
    return inval.split(delim);
}
exports.getDelimitedInput = getDelimitedInput;
function filePathSupplied(name) {
    // normalize paths
    var pathValue = path.resolve(this.getPathInput(name) || '');
    var repoRoot = path.resolve(this.getVariable('build.sourcesDirectory') || '');
    var supplied = pathValue !== repoRoot;
    debug(name + 'path supplied :' + supplied);
    return supplied;
}
exports.filePathSupplied = filePathSupplied;
function getPathInput(name, required, check) {
    var inval = getInput(name, required);
    if (inval) {
        if (check) {
            checkPath(inval, name);
        }
        if (inval.indexOf(' ') > 0) {
            if (!startsWith(inval, '"')) {
                inval = '"' + inval;
            }
            if (!endsWith(inval, '"')) {
                inval += '"';
            }
        }
    }
    debug(name + '=' + inval);
    return inval;
}
exports.getPathInput = getPathInput;
//-----------------------------------------------------
// Endpoint Helpers
//-----------------------------------------------------
function getEndpointUrl(id, optional) {
    var urlval = process.env['ENDPOINT_URL_' + id];
    if (!optional && !urlval) {
        _writeError('Endpoint not present: ' + id);
        exit(1);
    }
    debug(id + '=' + urlval);
    return urlval;
}
exports.getEndpointUrl = getEndpointUrl;
function getEndpointAuthorization(id, optional) {
    var aval = process.env['ENDPOINT_AUTH_' + id];
    if (!optional && !aval) {
        setResult(TaskResult.Failed, 'Endpoint not present: ' + id);
    }
    debug(id + '=' + aval);
    var auth;
    try {
        auth = JSON.parse(aval);
    }
    catch (err) {
        setResult(TaskResult.Failed, 'Invalid endpoint auth: ' + aval); // exit
    }
    return auth;
}
exports.getEndpointAuthorization = getEndpointAuthorization;
function stats(path) {
    return fs.statSync(path);
}
exports.stats = stats;
//-----------------------------------------------------
// Cmd Helpers
//-----------------------------------------------------
function command(command, properties, message) {
    var taskCmd = new tcm.TaskCommand(command, properties, message);
    _writeLine(taskCmd.toString());
}
exports.command = command;
function warning(message) {
    command('task.issue', { 'type': 'warning' }, message);
}
exports.warning = warning;
function error(message) {
    command('task.issue', { 'type': 'error' }, message);
}
exports.error = error;
function debug(message) {
    command('task.debug', null, message);
}
exports.debug = debug;
var _argStringToArray = function (argString) {
    var args = argString.match(/([^" ]*("[^"]*")[^" ]*)|[^" ]+/g);
    for (var i = 0; i < args.length; i++) {
        args[i] = args[i].replace(/"/g, "");
    }
    return args;
};
function cd(path) {
    if (path) {
        shell.cd(path);
    }
}
exports.cd = cd;
function pushd(path) {
    shell.pushd(path);
}
exports.pushd = pushd;
function popd() {
    shell.popd();
}
exports.popd = popd;
//------------------------------------------------
// Validation Helpers
//------------------------------------------------
function checkPath(p, name) {
    debug('check path : ' + p);
    if (!p || !fs.existsSync(p)) {
        setResult(TaskResult.Failed, 'not found ' + name + ': ' + p); // exit
    }
}
exports.checkPath = checkPath;
//-----------------------------------------------------
// Shell/File I/O Helpers
// Abstract these away so we can
// - default to good error handling
// - inject system.debug info
// - have option to switch internal impl (shelljs now)
//-----------------------------------------------------
function mkdirP(p) {
    var success = true;
    try {
        if (!p) {
            throw new Error('path not supplied');
        }
        // certain chars like \0 will cause shelljs and fs
        // to blow up without exception or error
        if (p.indexOf('\0') >= 0) {
            throw new Error('path cannot contain null bytes');
        }
        if (!shell.test('-d', p)) {
            debug('creating path: ' + p);
            shell.mkdir('-p', p);
            var errMsg = shell.error();
            if (errMsg) {
                handlerError(errMsg, false);
                success = false;
            }
        }
        else {
            debug('path exists: ' + p);
        }
    }
    catch (err) {
        success = false;
        handlerError('Failed mkdirP: ' + err.message, false);
    }
    return success;
}
exports.mkdirP = mkdirP;
function which(tool, check) {
    try {
        var toolPath = shell.which(tool);
        if (check) {
            checkPath(toolPath, tool);
        }
        debug(tool + '=' + toolPath);
        return toolPath;
    }
    catch (err) {
        handlerError('Failed which: ' + err.message, false);
    }
}
exports.which = which;
function cp(options, source, dest, continueOnError) {
    var success = true;
    try {
        shell.cp(options, source, dest);
        var errMsg = shell.error();
        if (errMsg) {
            handlerError(errMsg, continueOnError);
            success = false;
        }
    }
    catch (err) {
        success = false;
        handlerError('Failed cp: ' + err.message, false);
    }
    return success;
}
exports.cp = cp;
function mv(source, dest, force, continueOnError) {
    var success = true;
    try {
        if (force) {
            shell.mv('-f', source, dest);
        }
        else {
            shell.mv(source, dest);
        }
        var errMsg = shell.error();
        if (errMsg) {
            handlerError(errMsg, continueOnError);
            success = false;
        }
    }
    catch (err) {
        success = false;
        handlerError('Failed mv: ' + err.message, false);
    }
    return success;
}
exports.mv = mv;
function find(findPath) {
    try {
        if (!shell.test('-e', findPath)) {
            return [];
        }
        var matches = shell.find(findPath);
        debug('find ' + findPath);
        debug(matches.length + ' matches.');
        return matches;
    }
    catch (err) {
        handlerError('Failed find: ' + err.message, false);
    }
}
exports.find = find;
function rmRF(path, continueOnError) {
    var success = true;
    try {
        debug('rm -rf ' + path);
        shell.rm('-rf', path);
        var errMsg = shell.error();
        // if you try to delete a file that doesn't exist, desired result is achieved
        // other errors are valid
        if (errMsg && !(errMsg.indexOf('ENOENT') === 0)) {
            handlerError(errMsg, continueOnError);
            success = false;
        }
    }
    catch (err) {
        success = false;
        handlerError('Failed rmRF: ' + err.message, false);
    }
    return success;
}
exports.rmRF = rmRF;
function glob(pattern) {
    debug('glob ' + pattern);
    var matches = globm.sync(pattern);
    debug('found ' + matches.length + ' matches');
    if (matches.length > 0) {
        var m = Math.min(matches.length, 10);
        debug('matches:');
        if (m == 10) {
            debug('listing first 10 matches as samples');
        }
        for (var i = 0; i < m; i++) {
            debug(matches[i]);
        }
    }
    return matches;
}
exports.glob = glob;
function globFirst(pattern) {
    debug('globFirst ' + pattern);
    var matches = glob(pattern);
    if (matches.length > 1) {
        warning('multiple workspace matches.  using first.');
    }
    debug('found ' + matches.length + ' matches');
    return matches[0];
}
exports.globFirst = globFirst;
//-----------------------------------------------------
// Exec convenience wrapper
//-----------------------------------------------------
function exec(tool, args, options) {
    var toolPath = which(tool, true);
    var tr = createToolRunner(toolPath);
    if (args) {
        tr.arg(args);
    }
    return tr.exec(options);
}
exports.exec = exec;
function execSync(tool, args, options) {
    var toolPath = which(tool, true);
    var tr = createToolRunner(toolPath);
    if (args) {
        tr.arg(args);
    }
    return tr.execSync(options);
}
exports.execSync = execSync;
function createToolRunner(tool) {
    var tr = new trm.ToolRunner(tool);
    tr.on('debug', function (message) {
        debug(message);
    });
    return tr;
}
exports.createToolRunner = createToolRunner;
//-----------------------------------------------------
// Matching helpers
//-----------------------------------------------------
function match(list, pattern, options) {
    return minimatch.match(list, pattern, options);
}
exports.match = match;
function filter(pattern, options) {
    return minimatch.filter(pattern, options);
}
exports.filter = filter;
//-----------------------------------------------------
// Test Publisher
//-----------------------------------------------------
var TestPublisher = (function () {
    function TestPublisher(testRunner) {
        this.testRunner = testRunner;
    }
    TestPublisher.prototype.publish = function (resultFiles, mergeResults, platform, config) {
        if (mergeResults == 'true') {
            _writeLine("Merging test results from multiple files to one test run is not supported on this version of build agent for OSX/Linux, each test result file will be published as a separate test run in VSO/TFS.");
        }
        var properties = {};
        properties['type'] = this.testRunner;
        if (platform) {
            properties['platform'] = platform;
        }
        if (config) {
            properties['config'] = config;
        }
        for (var i = 0; i < resultFiles.length; i++) {
            command('results.publish', properties, resultFiles[i]);
        }
    };
    return TestPublisher;
})();
exports.TestPublisher = TestPublisher;
//-----------------------------------------------------
// Tools
//-----------------------------------------------------
exports.TaskCommand = tcm.TaskCommand;
exports.commandFromString = tcm.commandFromString;
exports.ToolRunner = trm.ToolRunner;
trm.debug = debug;
